### ISSUE TITLE
Retraining blender models with updated dataset

### ISSUE THREAD

**Issue by priyanksonis** [facebookresearch/ParlAI#2838]
Hi Team,
I want to retrain the blender with updated data (ex. On being asked who is president of United States ?, it should not tell "Barack Obama" and should tell instead "Donald Trump" ), because according to paper they used data through July 2019. I did not find the reddit dataset (from [https://files.pushshift.io/reddit/](url) as mentioned in paper) in list of tasks, Can you please help with how to download that dataset for needed time span followed by preprocessing (those 9 points in section 6.1 of [paper](https://arxiv.org/pdf/2004.13637.pdf) ) as mentioned in the paper?, I checked #2467, which is for a particular subreddit(eli5) but was not clear about using this for blender pretraining.

**Comment by stephenroller** [comment]
The pushshift resource only goes up to 2019-12, so that would be the most up to date you can get. (When we downloaded the data, 2019-07 was the most recent file.) https://files.pushshift.io/reddit/comments/RC_2019-12.zst would be the last month of comments, and you'll fine similarly named filenames for as far back as 2005.

We haven't released the pushshift preprocessing because it depends _heavily_ on our internal infrastructure, making it difficult to open source. But the paper describes how it's processed.

Here's how the data looks like after preprocessing, so you can get a feel of the format:
```
- - - NEW EPISODE: internal:new_reddit - - -
Dang how old is that bread?
   Haha not new but my gf let it get squished under some stuff :( I didnt want to put pants on to go to the store (which is why I made my own chips lol)
- - - NEW EPISODE: internal:new_reddit - - -
PSN: spagtwo  Always looking for new opponents to spar with. I'm a bit late, hope that's okay. My timezone is GMT but I can pretty much online at all hours of the day
That’s ok, I’m also GMT. Busy this evening (seeing extended family) but tomorrow is good?
   Hopefully tomorrow night.
- - - NEW EPISODE: internal:new_reddit - - -
Ég held að þau séu erlendis mest allt sumarið, en ég get spurt gítarleikarann fyrir þig. :)
   Litla fokking Ísland :)
- - - NEW EPISODE: internal:new_reddit - - -
The Fifth Season by N.K. Jemisin for sure. Weird fantastic setting, bizarre god-like races, interesting magic...unique and splendid.
Is it better than Inheritance
   I haven't read it yet but general consensus agrees that it's her best work to date. I very much enjoyed the Inheritance trilogy so I'm definitely excited to read this one.
- - - NEW EPISODE: internal:new_reddit - - -
Never settle ladies, you deserve the top 5 percentile ignore what those nerds keep saying about math.
stupid nerds, what have they ever done besides kick a football around?
   Or invent vibrators
- - - NEW EPISODE: internal:new_reddit - - -
First play from scrimmage, Luck picked off by Aqib Talib. He runs it back for 6. Please?
   TWIST- He runs the wrong way, fumbles, Reggie Wayne recovers it in our end zone.
- - - NEW EPISODE: internal:new_reddit - - -
Still parks better than half the people that live here it seems
I am one of the few people here, it seems, that cares about their car. I park in the furthest spot possible and still come out to find a jackass parked like a god damn potato parked their car.
Is no potato, only idiot in gas-donkey.
That sounds more like a fake Mexican accent.
I read it in a Russian accent.
ich beter yeh?
   Tad too much German, less Christoph Waltz, more homeless person drunk on vodka. And I mean that in the most respectful way, god I wish I was Russian.
- - - NEW EPISODE: internal:new_reddit - - -
I never should have written this article. Now, every time the net neutrality issue gets in the news I get a bunch of hate mail and requests to be on news programs because I'm the only guy they can find who's willing to go on the record against it.
   given that this article is a few years old, has your opinion on the matter changed at all?
- - - NEW EPISODE: internal:new_reddit - - -
Why are we qualifying McCarron ?   There must be something we dont see
There isn't. Bergevin just prefers size over speed.
we are one of the smallest fastest teams
Paul Byron and Victor Mete don't make the whole team fast.
   Drouin, Shaw, Gallagher, Domi, Lehkonen, Hudon, Danault. Patch is big but he’s still fast. Plekanec is honestly still fast for his age. DLR is very fast for his size. I just truly can’t comprehend how you came to this opinion.
- - - NEW EPISODE: internal:new_reddit - - -
ctrl+z
Oh man. This is embarrassing. Iunno why i didn't think about this first. I've been moonwalking for hours to no avail. I'll give this a try.
   No worries, happened to me last week ended up in a tub of peanut butter.
```

**Comment by priyanksonis** [comment]
Okay Stephen. 

In order to prepare updated dataset when it's released on [https://files.pushshift.io/reddit/comments/](url)

1. Do we need to first collect all the comments for a particular post with the help of `link_id` and sorting with `created_utc` ?
2. Then those 9 steps in section 6.1 of the paper.
3. Get all the data in files , so that we can use `-t fromfile:parlaiformat --fromfile_datapath` ?

Can you please correct me if am wrong ?

**Comment by stephenroller** [comment]
1. Download all the comments. No need to filter by link_id. You'll need to convert them from a tree structure ("flatten") the tree
2. Filter according to section 6.1 of the paper
3. I recommend using a ChunkTeacher rather than parlai format. The dump ends up being hundreds of gigabytes. You can find an example of a chunk teacher [here](https://github.com/facebookresearch/ParlAI/blob/1a65ad04bc2c05aad048a1dc9a3d6791aa2286e4/parlai/tasks/wikipedia/agents.py#L81).

Fair warning: we used 256 gpus to train BlenderBot.

**Comment by priyanksonis** [comment]
1. I am still not clear on this,  So data should be created like: "take each pair of successive comments in a thread to form the context and response of an example" (like mentioned in section 3.3 of this paper [Training Millions of Personalized Dialogue Agents](https://arxiv.org/pdf/1809.01984.pdf)) ?
as per the above paper, if we consider each pair then the sentence which was response will become context in other example. is it the way we do ?

I am referring this paper because:
blender paper  [Recipes for building an open-domain chatbot](https://arxiv.org/pdf/2004.13637.pdf) refers in sec. 6.1 --->  [Poly-encoders: architectures and pre-training
strategies for fast and accurate multi-sentence scoring](https://arxiv.org/pdf/1905.01969.pdf) refers in sec. 4.1 ---> [Training Millions of Personalized Dialogue Agents](https://arxiv.org/pdf/1809.01984.pdf)

**Comment by stephenroller** [comment]
If the tree looks like this:
```
A
  B
    C
  D
    E
```

We create 4 training examples (context -> label):
```
A -> B
AB -> C
A -> D
AD -> E
```

**Comment by priyanksonis** [comment]
Okay Thank you.
What json fields you used for getting the tree again from comments ? I tried with `link_id` , that gives all comments for a given post, but not sure that how can we arrange those comments to create original tree ?

**Comment by stephenroller** [comment]
IIRC:

- First pass: hash `link_id` in order to identify a chunk that each comment belongs to. We dump all comments with the same link_id hash into the same chunk. The number of chunks is chosen such that each chunk easily fits in memory.
- Second pass: for each chunk, load it fully into memory. Use the comment ID as a key and the full json structure as a value, I think. Use the `parent_id` field in each comment to set up the tree structure. Nodes with no parent_id (I think) become the roots. All others get linked
- Once the tree structure is created, do an pre-order traversal, keeping track of the path to the root. Then at every node, use the node and this path to root to construct the dialogue history.

**Comment by priyanksonis** [comment]
ok, @stephenroller 

These filters in sec 6.1 of [paper](https://arxiv.org/pdf/2004.13637.pdf), how do you check for them:

1. The author is a known bot.
2. It comes from a known non-English subreddit.
5. It is longer than 128 BPE tokens.

-For first one I thought of using where author name contains word : `"bot"`. but does not seems right way. do you have any list of known bots or something ?
-For last one can we use dictionary `"model.dict"` given in `"data/models/blender/blender_3B"` ?, but needs more clarity on how we can get BPE tokens for a given string ?

**Comment by stephenroller** [comment]
- On bot: This is the exact heuristic we used
- We use HF's tokenizers to tokenize the text and count
- On non-English: here's a list of subreddits we filtered:

      mexico ateosmexicanos cinefilos explicacomoanino futbolmx futbol tigres
      rayados mejorde mexicocirclejerk mexicoiama mexijobs nahuatl nerdsmexico
      tvadicto videojuego gamingesp aguascalientes bajacalifornia tijuana
      mexicali ensenada rosarito distritofederal coyoacan aguascalientes
      bajacaliforniasur guanajuato estadodemexico oaxaca queretaro quintanaroo
      sinaloa sonora veracruz yucatan zacatecas edomex guadalajara guanajuato
      lalaguna monterrey playa es redditores argentina argentinacirclejerk
      argentos espanol bolivia chile santiago ecuador elsalvador latinoamerica
      paraguay peru colombia guatemala honduras nicaragua panama puertorico
      uruguay uruguaysoccer vzla arteperuano programacion djangoes role practicar
      madrid chistes noticias 15m valencia simpsonsespanollatino spanishimmersion
      france bretagne grandsud aixmarseille provence bordeaux montpellier
      toulouse lille lyon nantes fr cameroon morocco quebec montreal etsmtl udem
      uqam polymtl ulaval udes photosmontreal montrealgay montrealjobs
      montrealfood r4rmontreal occuponsmontreal occuponsmontrealag
      occuponsmontrealjobs spvm ecouteca entraideinformatique frenchrap jememarre
      jeuxvideo livres ligue1 programmation truefrance ragefrance lecafe
      demandezreddit frenchimmersion portuguese portugal porto aveiro coimbra
      lisboa desporto francesinhas benfica reddist brasil saopaulo riograndedosul
      brasilia bbrasil bossanova italy torino genova venezia napoli milano
      firenze italians erba soluzioni programmazione de deutschland frankfurt
      koeln germanpractice austria linz wien atomausstieg dokumentationen
      de_comedy germusic aachen germanmovies bern niedersachsen schleswigholstein
      hannover darmstadt bremen cologne hamburg frankfurt freiburg piratenpartei
      datenschutz de_it fragreddit teutonik germanyusa fernsehen hoerbuecher
      bundesliga heidelberg goe germantrees nederlands vlaanderen groningen
      nijmegen nl nederland antwerpen gent sweden svenska bostad jobb marknad
      swiama swadviceanimals ilandsproblem swedents swirclejerk metaswirclejerk
      stockholm malmo lund gothenburg uppsala linkoping umea skellefte gavle
      orebro vasteras karlstad boras halmstad denmark dkpol danishents dkmusik
      danskrap dkwork aalborg aarhus kolding odense norge oslo ntnu trondheim
      bergen stavanger polska polityka muzyka poznan lodz krakow ru
      russianimmersion suomi helsinki tampere chinesereddit cpop cflcomics
      japanesereddit nihongo nihon japaneseonly ja eli5
